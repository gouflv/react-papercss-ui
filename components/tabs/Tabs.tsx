import { useUpdateEffect } from 'ahooks'
import classNames from 'classnames'
import React, {
  FC,
  PropsWithChildren,
  ReactElement,
  ReactNode,
  useEffect,
  useState
} from 'react'
import { toArray } from '../../utils/children'

export type TabsProps = PropsWithChildren<{
  activeKey?: string | number
  onChange?: (value) => void
}>

const Tabs = (props: PropsWithChildren<TabsProps>) => {
  const [active, setActive] = useState<string>()

  useEffect(() => {
    if (!props.activeKey) return
    setActive(String(props.activeKey))
  }, [props.activeKey])

  function tabNavRender() {
    return toArray<ReactElement>(props.children).map((child, index) => {
      const key = String(child.props.value) || String(index)
      return (
        <label
          key={key}
          className={classNames({
            active: key === active
          })}
          onClick={() => setActive(key)}
        >
          {child.props.tab}
        </label>
      )
    })
  }

  function contentRender() {
    return toArray<ReactElement>(props.children).map((child, index) => {
      const key = String(child.props.value) || String(index)
      const isActive = active === key
      return React.cloneElement(child, {
        ...child.props,
        value: key,
        active: isActive
      } as TabPaneProps)
    })
  }

  useUpdateEffect(() => {
    if (typeof props.onChange === 'function' && active !== props.activeKey) {
      props.onChange(active)
    }
  }, [active])

  return (
    <div className='tabs'>
      <div className='tabs-nav'>{tabNavRender()}</div>
      <div className='tabs-content'>{contentRender()}</div>
    </div>
  )
}

export interface TabPaneProps {
  tab: ReactNode
  value?: number | string
  active?: boolean
}

const TabPane: FC<TabPaneProps> = props => {
  return (
    <div
      className={classNames('content', {
        active: props.active
      })}
    >
      {props.children}
    </div>
  )
}

Tabs.TabPane = TabPane

export { Tabs }
