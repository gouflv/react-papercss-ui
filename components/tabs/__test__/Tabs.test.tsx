import '@testing-library/jest-dom'
import { fireEvent, render, screen } from '@testing-library/react'
import * as React from 'react'
import { PropsWithChildren } from 'react'
import { Tabs, TabsProps } from '../Tabs'

const { TabPane } = Tabs

describe('Test Panel Component', () => {
  let props: PropsWithChildren<TabsProps>

  beforeEach(() => {
    props = {
      children: [
        <TabPane key='1' tab='First Tab' value='1'>
          First Tab Content
        </TabPane>,
        <TabPane key='2' tab='Second Tab' value='2'>
          Second Tab Content
        </TabPane>
      ]
    }
  })

  const renderComponent = () => render(<Tabs {...props} />)

  it('should open panel without active setting', () => {
    const { container } = renderComponent()
    expect(container).toMatchSnapshot()
    expect(container.querySelector('.active')).toBeNull()
  })

  it('should open panel with active set as "2"', () => {
    props = {
      ...props,
      activeKey: '2'
    }
    const { container } = renderComponent()
    expect(container).toMatchSnapshot()

    const tabsNav = screen.getByText('Second Tab')
    expect(tabsNav).toHaveClass('active')

    const tabsContent = screen.getByText('Second Tab Content')
    expect(tabsContent).toHaveClass('active')
  })

  it('should emit active key on click tab-nav', () => {
    const onChange = jest.fn()
    props = {
      ...props,
      activeKey: '1',
      onChange
    }
    const cmp = renderComponent()
    fireEvent.click(cmp.getByText('Second Tab'))
    expect(onChange).toBeCalledWith('2')
    expect(onChange).toBeCalledTimes(1)
  })

  it('should open un-name panel with active set as number', () => {
    props = {
      activeKey: 1,
      children: [
        <TabPane key='1' tab='First Tab'>
          First Tab Content
        </TabPane>,
        <TabPane key='2' tab='Second Tab'>
          Second Tab Content
        </TabPane>
      ]
    }

    const { container } = renderComponent()
    expect(container).toMatchSnapshot()
    const tabsNav = screen.getByText('Second Tab')
    expect(tabsNav).toHaveClass('active')

    const tabsContent = screen.getByText('Second Tab Content')
    expect(tabsContent).toHaveClass('active')
  })
})
