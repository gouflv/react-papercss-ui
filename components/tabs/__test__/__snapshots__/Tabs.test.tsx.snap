// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Panel Component should open panel with active set as "2" 1`] = `
<div>
  <div
    class="tabs"
  >
    <div
      class="tabs-nav"
    >
      <label
        class=""
      >
        First Tab
      </label>
      <label
        class="active"
      >
        Second Tab
      </label>
    </div>
    <div
      class="tabs-content"
    >
      <div
        class="content"
      >
        First Tab Content
      </div>
      <div
        class="content active"
      >
        Second Tab Content
      </div>
    </div>
  </div>
</div>
`;

exports[`Test Panel Component should open panel without active setting 1`] = `
<div>
  <div
    class="tabs"
  >
    <div
      class="tabs-nav"
    >
      <label
        class=""
      >
        First Tab
      </label>
      <label
        class=""
      >
        Second Tab
      </label>
    </div>
    <div
      class="tabs-content"
    >
      <div
        class="content"
      >
        First Tab Content
      </div>
      <div
        class="content"
      >
        Second Tab Content
      </div>
    </div>
  </div>
</div>
`;

exports[`Test Panel Component should open un-name panel with active set as number 1`] = `
<div>
  <div
    class="tabs"
  >
    <div
      class="tabs-nav"
    >
      <label
        class=""
      >
        First Tab
      </label>
      <label
        class="active"
      >
        Second Tab
      </label>
    </div>
    <div
      class="tabs-content"
    >
      <div
        class="content"
      >
        First Tab Content
      </div>
      <div
        class="content active"
      >
        Second Tab Content
      </div>
    </div>
  </div>
</div>
`;
